feat: add resource profiler script

rickstaa opened the pull request:
This pull request adds a small profiler script developers can use while working on ComfyStream to check resource usage and potential bottlenecks. It will provide the developer with the following information:

<details>
<summary>Show Data</summary>

```bash
(comfystream) [16:26:29] ricks@rick-dev-station:~/development/livepeer/ai/comfystream_ws/comfystream/scripts$ sudo env "PATH=$PATH" python monitor_pid_resources.py --pid  68618 --output resources_output.csv --duration 120 --spy
Monitoring system resources for PID 68618 for 120 seconds...
Running py-spy for deep profiling...
CPU: 298.50%, RAM: 5512.73MB, GPU: 44.00%, VRAM: 3876.00MB
CPU: 299.50%, RAM: 5512.73MB, GPU: 28.00%, VRAM: 3876.00MB
CPU: 296.50%, RAM: 5512.73MB, GPU: 49.00%, VRAM: 3876.00MB
CPU: 307.50%, RAM: 5512.98MB, GPU: 41.00%, VRAM: 3876.00MB
CPU: 302.00%, RAM: 5512.98MB, GPU: 41.00%, VRAM: 3876.00MB
CPU: 307.00%, RAM: 5512.98MB, GPU: 35.00%, VRAM: 3876.00MB
CPU: 268.00%, RAM: 5512.98MB, GPU: 38.00%, VRAM: 3876.00MB
CPU: 298.50%, RAM: 5512.98MB, GPU: 43.00%, VRAM: 3876.00MB
CPU: 299.50%, RAM: 5512.98MB, GPU: 33.00%, VRAM: 3876.00MB
CPU: 305.50%, RAM: 5512.98MB, GPU: 34.00%, VRAM: 3876.00MB
CPU: 291.00%, RAM: 5512.98MB, GPU: 36.00%, VRAM: 3876.00MB
CPU: 299.00%, RAM: 5512.98MB, GPU: 43.00%, VRAM: 3876.00MB
CPU: 295.50%, RAM: 5535.23MB, GPU: 35.00%, VRAM: 3876.00MB
CPU: 283.50%, RAM: 5554.73MB, GPU: 36.00%, VRAM: 3876.00MB
CPU: 283.50%, RAM: 5569.48MB, GPU: 35.00%, VRAM: 3876.00MB
CPU: 286.50%, RAM: 5597.48MB, GPU: 38.00%, VRAM: 3876.00MB
CPU: 292.50%, RAM: 5603.48MB, GPU: 38.00%, VRAM: 3876.00MB
CPU: 283.00%, RAM: 5625.73MB, GPU: 38.00%, VRAM: 3876.00MB
CPU: 316.00%, RAM: 5640.23MB, GPU: 39.00%, VRAM: 3876.00MB
CPU: 278.00%, RAM: 5665.48MB, GPU: 38.00%, VRAM: 3876.00MB
CPU: 299.50%, RAM: 5676.73MB, GPU: 43.00%, VRAM: 3876.00MB
CPU: 296.00%, RAM: 5701.98MB, GPU: 32.00%, VRAM: 3876.00MB
CPU: 300.00%, RAM: 5718.98MB, GPU: 41.00%, VRAM: 3876.00MB
CPU: 299.50%, RAM: 5730.23MB, GPU: 37.00%, VRAM: 3876.00MB
CPU: 316.00%, RAM: 5746.98MB, GPU: 43.00%, VRAM: 3876.00MB
CPU: 306.50%, RAM: 5766.48MB, GPU: 40.00%, VRAM: 3876.00MB
CPU: 299.50%, RAM: 5792.23MB, GPU: 42.00%, VRAM: 3876.00MB
CPU: 286.00%, RAM: 5811.98MB, GPU: 38.00%, VRAM: 3876.00MB
CPU: 279.50%, RAM: 5825.98MB, GPU: 32.00%, VRAM: 3876.00MB
CPU: 309.00%, RAM: 5849.73MB, GPU: 39.00%, VRAM: 3876.00MB
CPU: 224.00%, RAM: 5858.98MB, GPU: 41.00%, VRAM: 3876.00MB
CPU: 312.00%, RAM: 5858.98MB, GPU: 43.00%, VRAM: 3876.00MB
CPU: 309.50%, RAM: 5858.98MB, GPU: 38.00%, VRAM: 3876.00MB
CPU: 302.50%, RAM: 5858.98MB, GPU: 34.00%, VRAM: 3876.00MB
CPU: 290.00%, RAM: 5858.98MB, GPU: 47.00%, VRAM: 3876.00MB
CPU: 298.50%, RAM: 5858.98MB, GPU: 45.00%, VRAM: 3876.00MB
CPU: 295.50%, RAM: 5858.98MB, GPU: 36.00%, VRAM: 3876.00MB
CPU: 300.00%, RAM: 5858.98MB, GPU: 42.00%, VRAM: 3876.00MB
CPU: 288.00%, RAM: 5858.98MB, GPU: 43.00%, VRAM: 3876.00MB
CPU: 297.00%, RAM: 5859.23MB, GPU: 31.00%, VRAM: 3876.00MB
CPU: 291.00%, RAM: 5859.23MB, GPU: 35.00%, VRAM: 3876.00MB
CPU: 298.50%, RAM: 5859.23MB, GPU: 37.00%, VRAM: 3876.00MB
CPU: 310.50%, RAM: 5859.23MB, GPU: 38.00%, VRAM: 3876.00MB
CPU: 309.50%, RAM: 5859.23MB, GPU: 40.00%, VRAM: 3876.00MB
CPU: 296.50%, RAM: 5859.48MB, GPU: 35.00%, VRAM: 3876.00MB
CPU: 311.00%, RAM: 5859.98MB, GPU: 37.00%, VRAM: 3876.00MB
CPU: 305.50%, RAM: 5859.98MB, GPU: 37.00%, VRAM: 3876.00MB
CPU: 299.00%, RAM: 5859.98MB, GPU: 42.00%, VRAM: 3876.00MB
CPU: 298.00%, RAM: 5862.23MB, GPU: 34.00%, VRAM: 3876.00MB
CPU: 281.50%, RAM: 5862.23MB, GPU: 30.00%, VRAM: 3876.00MB
CPU: 292.50%, RAM: 5862.48MB, GPU: 38.00%, VRAM: 3876.00MB
CPU: 305.00%, RAM: 5862.48MB, GPU: 48.00%, VRAM: 3876.00MB
CPU: 306.00%, RAM: 5862.48MB, GPU: 37.00%, VRAM: 3876.00MB
CPU: 306.00%, RAM: 5865.23MB, GPU: 36.00%, VRAM: 3876.00MB
CPU: 273.00%, RAM: 5865.23MB, GPU: 37.00%, VRAM: 3876.00MB
CPU: 318.50%, RAM: 5865.48MB, GPU: 37.00%, VRAM: 3876.00MB
CPU: 308.00%, RAM: 5865.73MB, GPU: 47.00%, VRAM: 3876.00MB
CPU: 321.50%, RAM: 5865.73MB, GPU: 40.00%, VRAM: 3876.00MB
CPU: 301.50%, RAM: 5865.73MB, GPU: 36.00%, VRAM: 3876.00MB
Py-Spy flame graph saved to pyspy_profile.svg
CPU: 410.00%, RAM: 5865.73MB, GPU: 56.00%, VRAM: 3876.00MB
AVERAGE - CPU: 299.04%, RAM: 5740.08MB, GPU: 38.77%, VRAM: 3876.00MB
Logs saved to resources_output.csv
```

[resources_output.csv](https://github.com/user-attachments/files/18911161/resources_output.csv)

![pyspy_profile](https://github.com/user-attachments/assets/1b3aa145-0dd5-41a9-a9f4-87f05b339b2c)

</summary>

eliteprox commented on scripts/monitor_pid_resources.py:
```diff
@@ -0,0 +1,155 @@
+            process_gpu_usage = 0
+            process_vram_usage = 0
+            device_count = pynvml.nvmlDeviceGetCount()
+            for i in range(device_count):
+                handle = pynvml.nvmlDeviceGetHandleByIndex(i)
+                processes = pynvml.nvmlDeviceGetComputeRunningProcesses(handle)
+                for proc_info in processes:
+                    if proc_info.pid == pid:
+                        process_gpu_usage = pynvml.nvmlDeviceGetUtilizationRates(
+                            handle
+                        ).gpu
+                        process_vram_usage = proc_info.usedGpuMemory / (
+                            1024 * 1024
+                        )  # MB
+                        break
```
Upon testing this with the latest hiddenswitch upgrade, the process id utilizing the GPU is constantly changing. (observed with `ps aux | grep app.py`). We could use the pynml library to find processes that are currently utilizing the GPU:

```
gpu_processes = []
try:
    handle = pynvml.nvmlDeviceGetHandleByIndex(i)
    processes = pynvml.nvmlDeviceGetComputeRunningProcesses(handle)
    for proc_info in processes:
        process_gpu_usage = pynvml.nvmlDeviceGetUtilizationRates(handle).gpu
        process_vram_usage = proc_info.usedGpuMemory / (1024 * 1024)  # MB
        if process_gpu_usage > 0:
            gpu_processes.append({
                "pid": proc_info.pid,
                "gpu_usage": process_gpu_usage,
                "vram_usage": process_vram_usage
            })
except pynvml.NVMLError as err:
    print(f"Failed to get GPU information: {err}")

# Log GPU processes utilization
for gpu_proc in gpu_processes:
    print(f"PID: {gpu_proc['pid']}, GPU Usage: {gpu_proc['gpu_usage']}%, VRAM Usage: {gpu_proc['vram_usage']} MB")
```

rickstaa replied:
@eliteprox okay this was fun but I got it working. I checked it by comparing it with the output of [atop](https://www.atoptool.nl/) (i.e. `atop 2` and hitting `p`).

eliteprox reviewed (changes requested):
Currently the script is inaccurate at measuring GPU usage due to how ComfyUI is launched with ProcessPoolExecutor. Let's improve the solution so vram and gpu utilization can be measured

eliteprox commented:
Still having issues getting the GPU utilization. Are you testing inside a devcontainer? Maybe this works better at the host level?

```
(comfystream) root@2b7fc23b9f3f:/workspace/comfystream/scripts# pgrep -f app.py | xargs ps -o pid,cmd --pid
    PID CMD
  34442 /workspace/miniconda3/envs/comfystream/bin/python -u /root/.vscode-server/extensions/ms-python.debugpy-2025.4.0-linux-x64/bundled/libs/debugpy/_vendored/pydevd/pydevd.py --port 39281 --ppid 33741 --client 127.0.0.1 --client-access-token 4679bee715a9125
(comfystream) root@2b7fc23b9f3f:/workspace/comfystream/scripts# python monitor_pid_resources.py --pid 34442 
Monitoring PID 34442 for 30 seconds...
[0.0%] CPU: 307.90%, RAM: 9419.47MB, GPU: 0.00%, VRAM: 0.00MB
[6.7%] CPU: 218.40%, RAM: 9419.47MB, GPU: 0.00%, VRAM: 0.00MB
[13.3%] CPU: 248.30%, RAM: 9419.47MB, GPU: 0.00%, VRAM: 0.00MB
[20.0%] CPU: 297.40%, RAM: 9419.47MB, GPU: 0.00%, VRAM: 0.00MB
[26.7%] CPU: 238.50%, RAM: 9419.59MB, GPU: 0.00%, VRAM: 0.00MB
```

rickstaa commented:
> Still having issues getting the GPU utilization. Are you testing inside a devcontainer? Maybe this works better at the host level?
> 
> ```
> (comfystream) root@2b7fc23b9f3f:/workspace/comfystream/scripts# pgrep -f app.py | xargs ps -o pid,cmd --pid
>     PID CMD
>   34442 /workspace/miniconda3/envs/comfystream/bin/python -u /root/.vscode-server/extensions/ms-python.debugpy-2025.4.0-linux-x64/bundled/libs/debugpy/_vendored/pydevd/pydevd.py --port 39281 --ppid 33741 --client 127.0.0.1 --client-access-token 4679bee715a9125
> (comfystream) root@2b7fc23b9f3f:/workspace/comfystream/scripts# python monitor_pid_resources.py --pid 34442 
> Monitoring PID 34442 for 30 seconds...
> [0.0%] CPU: 307.90%, RAM: 9419.47MB, GPU: 0.00%, VRAM: 0.00MB
> [6.7%] CPU: 218.40%, RAM: 9419.47MB, GPU: 0.00%, VRAM: 0.00MB
> [13.3%] CPU: 248.30%, RAM: 9419.47MB, GPU: 0.00%, VRAM: 0.00MB
> [20.0%] CPU: 297.40%, RAM: 9419.47MB, GPU: 0.00%, VRAM: 0.00MB
> [26.7%] CPU: 238.50%, RAM: 9419.59MB, GPU: 0.00%, VRAM: 0.00MB
> ```

@eliteprox Thanks for catching that! The script is intended to be used at the host level because `pynvml` relies on host PIDs, which containers don’t have direct access to. While users could try `--pid=host`, this often leads to permission issues with the `psutil` package. To address this, I’ve updated the script to display a warning and added a `--host-pid` option, allowing users to run it inside a container while still being able to obtain average GPU usage.

eliteprox reviewed (approved):
LGTM!
